### PR TITLE
Fix unhandled UnsupportedOperationException in Fallocate

### PR DESCRIPTION
**Pull request by torusrxxx**

### CODE REVIEW

File: src/freenet/support/io/Fallocate.java [hunk 1]
```diff
@@ -19,121 +19,138 @@
...
+				errno = result == 0 ? 0 : Native.getLastError();
+			} else if (IS_POSIX) {
+				errno = FallocateHolderPOSIX.posix_fallocate(fd, offset, final_filesize-offset);
+			} else {
+				isUnsupported = true;
+			}
+		} else {
+			if (Platform.isWindows()) {
+				// On Windows, just write a byte at the end.
+				channel.write(ByteBuffer.allocate(1), final_filesize - 1);
```
Comment (ArneBab) [inline src/freenet/support/io/Fallocate.java:93]: why is this safe?

File: src/freenet/support/io/Fallocate.java [hunk 2]
```diff
@@ -19,121 +19,138 @@
...
+		this.fd = fd;
+		this.final_filesize = final_filesize;
+		this.channel = channel;
+	}
+
+	public static Fallocate forChannel(FileChannel channel, long final_filesize) {
+		try {
+			return new Fallocate(channel, getDescriptor(channel), final_filesize);
+		} catch (final UnsupportedOperationException exception) {
+			return new Fallocate(channel, 0, final_filesize);
```
Comment (ArneBab) [inline src/freenet/support/io/Fallocate.java:44]: It would be good to have a comment here that explains why 0 is the right file descriptor.

**Comment by ArneBab** [comment]
Please stick to the original indentation style. Re-indenting the file breaks annotations to trace changes to commits.

That said: Thank you for working on this!

**Comment by torusrxxx** [comment]
Recreated in #1010

**Pull request closed without merging**